pkg/controller/eventing/inmemory/channel/reconcile.go

```diff
@@ -0,0 +1,450 @@
+	c.Status.Conditions = []duckv1alpha1.Condition{
+		{
+			Type:   duckv1alpha1.ConditionReady,
+			Status: corev1.ConditionTrue,
+		},
+	}
+}
+func (r *reconciler) updateChannel(ctx context.Context, u *eventingv1alpha1.Channel) error {
+	o := &eventingv1alpha1.Channel{}
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: u.Namespace, Name: u.Name}, o)
+	if err != nil {
+		r.logger.Info("Error getting Channel for status update", zap.Error(err), zap.Any("updatedChannel", u))
+		return err
+	}
+
+	updated := false
+	if !equality.Semantic.DeepEqual(o.Finalizers, u.Finalizers) {
+		updated = true
+		o.SetFinalizers(u.Finalizers)
+	}
+	if !equality.Semantic.DeepEqual(o.Status, u.Status) {
+		updated = true
+		o.Status = u.Status
+	}
+
+	if updated {
+		return r.client.Update(ctx, o)
+	}
+	return nil
+}
+
+func (r *reconciler) syncChannelConfig(ctx context.Context) error {
+	channels, err := r.listAllChannels(ctx)
+	if err != nil {
+		r.logger.Info("Unable to list channels", zap.Error(err))
+		return err
+	}
+	config := multiChannelFanoutConfig(channels)
+	return r.writeConfigMap(ctx, config)
+}
+
+func (r *reconciler) writeConfigMap(ctx context.Context, config *multichannelfanout.Config) error {
+	logger := r.logger.With(zap.Any("configMap", r.configMapKey))
+
+	updated, err := multichannelfanoutparse.SerializeConfig(*config)
+	if err != nil {
+		r.logger.Error("Unable to serialize config", zap.Error(err), zap.Any("config", config))
+		return err
+	}
+
+	cm := &corev1.ConfigMap{}
+	err = r.client.Get(ctx, r.configMapKey, cm)
+	if errors.IsNotFound(err) {
+		cm = r.createNewConfigMap(updated)
+		err = r.client.Create(ctx, cm)
+	}
+	if err != nil {
+		logger.Info("Unable to get/create ConfigMap", zap.Error(err))
+		return err
+	}
+
+	cur := cm.Data[multichannelfanoutparse.MultiChannelFanoutConfigKey]
+	if equality.Semantic.DeepEqual(cur, updated) {
+		// Nothing to update.
+		return nil
+	}
+
+	cm.Data[multichannelfanoutparse.MultiChannelFanoutConfigKey] = updated
+	return r.client.Update(ctx, cm)
+}
+
+func (r *reconciler) createNewConfigMap(config string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.configMapKey.Namespace,
+			Name:      r.configMapKey.Name,
+		},
+		Data: map[string]string{
+			multichannelfanoutparse.MultiChannelFanoutConfigKey: config,
+		},
+	}
+}
+
+func multiChannelFanoutConfig(channels []eventingv1alpha1.Channel) *multichannelfanout.Config {
+	cc := make([]multichannelfanout.ChannelConfig, 0)
+	for _, c := range channels {
+		channelable := c.Spec.Channelable
+		if channelable != nil {
+			cc = append(cc, multichannelfanout.ChannelConfig{
+				Namespace: c.Namespace,
+				Name:      c.Name,
+				FanoutConfig: fanout.Config{
+					Subscriptions: c.Spec.Channelable.Subscribers,
+				},
+			})
+		}
+	}
+	return &multichannelfanout.Config{
+		ChannelConfigs: cc,
+	}
+}
+
+func (r *reconciler) listAllChannels(ctx context.Context) ([]eventingv1alpha1.Channel, error) {
```

Comment:
would be better to use an informer lister so that a cache can be preserved and maintained between calls rather than DOSing the api server with requests.

Reply:
My assumption is that the controller runtime, which creates the client, is doing the caching by itself. I will verify this assumption before submitting.

Reply:
@adamharwayne is correct; this will always be a cached read. The controller-runtime client automatically manages informers for all types and delegates reads to their caches. This is not well documented but here's the caching client interface: https://github.com/kubernetes-sigs/controller-runtime/blob/ddf0390deebe496cb449b1b8540031da9236afa2/pkg/client/split.go

and its use by the manager: https://github.com/kubernetes-sigs/controller-runtime/blob/53fc44b56078cd095b11bd44cfa0288ee4cf718f/pkg/manager/manager.go#L197